### PR TITLE
pallet-asset-conversion: Decoupling Native Currency Dependancy

### PR DESCRIPTION
**Pull request by muharem**
closes https://github.com/paritytech/polkadot-sdk/issues/1842

Decoupling Pallet from the Concept of Native Currency

Currently, the pallet is intrinsically linked with the concept of native currency, requiring users to provide implementations of the `fungible::*` and `fungibles::*` traits to interact with native and non native assets. This incapsulates some non-related to the pallet complexity and makes it less adaptable in contexts where the native currency concept is absent.

With this PR, the dependence on `fungible::*` for liquidity-supplying assets has been removed. Instead, the native and non-native currencies' handling is now overseen by a single type that implements the `fungibles::*` traits. To simplify this integration, types have been introduced to facilitate the creation of a union between `fungible::*` and `fungibles::*` implementations, producing a unified `fungibles::*` type.

One of the reasons driving these changes is the ambition to create a more user-friendly API for the `SwapCredit` implementation. Given that it interacts with two distinct credit types from `fungible` and `fungibles`, a unified type was introduced. Clients now manage potential conversion failures for those credit types. In certain contexts, it's vital to guarantee that operations are fail-safe, like in this impl - [PR](https://github.com/paritytech/polkadot-sdk/pull/1845), place in [code](https://github.com/paritytech/polkadot-sdk/blob/20b85a5fada8f55c98ba831964f5866ffeadf4da/cumulus/primitives/utility/src/lib.rs#L429).

Additional Updates:
- abstracted the pool ID and its account derivation logic via trait bounds, along with common implementation offerings;
- removed `inc_providers` on a pool creation for the pool account;
- benchmarks:
-- swap complexity is N, not const;
-- removed `From<u128> + Into<u128>` bound from `T::Balance`;
-- removed swap/liquidity/.. amount constants, resolve them dynamically based on pallet configuration;
-- migrated to v2 API;
- `OnUnbalanced` handler for the pool creation fee, replacing direct transfers to a specified account ID;
- renamed `MultiAssetId` to `AssetKind` aligning with naming across frame crates;

related PRs:
- (depends) https://github.com/paritytech/polkadot-sdk/pull/1677
- (caused) https://github.com/paritytech/polkadot-sdk/pull/2033
- (caused) https://github.com/paritytech/polkadot-sdk/pull/1876

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4085556

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4130304

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4130979

File: cumulus/parachains/runtimes/assets/asset-hub-kusama/src/weights/pallet_asset_conversion.rs [hunk 1]
```diff
@@ -66,22 +62,22 @@ impl<T: frame_system::Config> pallet_asset_conversion::WeightInfo for WeightInfo
...
-		//  Measured:  `480`
-		//  Estimated: `6196`
-		// Minimum execution time: 88_484_000 picoseconds.
-		Weight::from_parts(92_964_000, 0)
-			.saturating_add(Weight::from_parts(0, 6196))
-			.saturating_add(T::DbWeight::get().reads(9))
+		//  Measured:  `408`
+		//  Estimated: `4689`
+		// Minimum execution time: 905_000_000 picoseconds.
+		Weight::from_parts(928_000_000, 0)
```
Comment (franciscoaguirre) [inline cumulus/parachains/runtimes/assets/asset-hub-kusama/src/weights/pallet_asset_conversion.rs:68]: Any idea why this went up ten times?

File: cumulus/parachains/runtimes/assets/asset-hub-kusama/src/weights/pallet_asset_conversion.rs [hunk 2]
```diff
@@ -90,34 +86,32 @@ impl<T: frame_system::Config> pallet_asset_conversion::WeightInfo for WeightInfo
 		// Proof Size summary in bytes:
 		//  Measured:  `1117`
 		//  Estimated: `7404`
-		// Minimum execution time: 153_015_000 picoseconds.
-		Weight::from_parts(157_018_000, 0)
+		// Minimum execution time: 1_628_000_000 picoseconds.
+		Weight::from_parts(1_799_000_000, 0)
```
Comment (franciscoaguirre) [inline cumulus/parachains/runtimes/assets/asset-hub-kusama/src/weights/pallet_asset_conversion.rs:90]: This shot up even more

File: substrate/frame/asset-conversion/src/tests.rs [hunk 3]
```diff
@@ -64,40 +66,71 @@ fn pool_assets() -> Vec<u32> {
...
 		vec![$( Box::new( $x ), )*]
-	}
+	)
+}
+
+#[test]
+fn validate_with_first_asset_pool_id_locator() {
+	new_test_ext().execute_with(|| {
+		use NativeOrWithId::{Native, WithId};
+		assert_eq!(WithFirstAssetLocator::pool_id(&Native, &WithId(2)), Ok((Native, WithId(2))));
```
Comment (jsidorenko) [inline substrate/frame/asset-conversion/src/tests.rs:105]: `WithFirstAssetLocator` - shouldn't it be 'WithFirst**Native**Locator'?

Comment (jsidorenko) [reply substrate/frame/asset-conversion/src/tests.rs:105]: ok, I see where it comes from. I would put a type alias like `type NativeFirstLocator = WithFirstAssetLocator<...>`

File: cumulus/parachains/runtimes/assets/asset-hub-kusama/src/weights/pallet_asset_conversion.rs (hunk 1 above)
Comment (muharem) [reply cumulus/parachains/runtimes/assets/asset-hub-kusama/src/weights/pallet_asset_conversion.rs:68]: probably my local machine, and target is `debug` instead `production`, you can see it in the header

**Review by franciscoaguirre — APPROVED** [review]
I'd use bench bot to regenerate weights properly before merging or immediately after

**Review by jsidorenko — APPROVED** [review]

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-6-0/5855/1

**Pull request merged**